File: apps/rust-server/src/services/pending_events.rs
use deadpool_redis::Pool;
use deadpool_redis::redis::AsyncCommands;
use serde_json::Value;

const KEY_PREFIX: &str = "pending_ws_events:";
const MAX_EVENTS_PER_USER: i64 = 1000;
const TTL_SECONDS: i64 = 86400; // 24 hours

fn key(user_id: &str) -> String {
    format!("{}{}", KEY_PREFIX, user_id)
}

/// Push a WS event to a user's pending queue.
/// Uses timestamp as score for ordering.
pub async fn push_event(redis: &Pool, user_id: &str, event: &Value) -> Result<(), anyhow::Error> {
    let mut conn = redis.get().await?;
    let k = key(user_id);
    let score = chrono::Utc::now().timestamp_millis() as f64;
    let event_str = serde_json::to_string(event)?;

    conn.zadd::<_, _, _, ()>(&k, &event_str, score).await?;

    // Cap at MAX_EVENTS_PER_USER (remove oldest)
    let count: i64 = conn.zcard(&k).await?;
    if count > MAX_EVENTS_PER_USER {
        conn.zremrangebyrank::<_, ()>(&k, 0, (count - MAX_EVENTS_PER_USER - 1) as isize).await?;
    }

    // Reset TTL
    conn.expire::<_, ()>(&k, TTL_SECONDS).await?;

    Ok(())
}

/// Get all pending events for a user.
pub async fn get_pending_events(redis: &Pool, user_id: &str) -> Result<Vec<Value>, anyhow::Error> {
    let mut conn = redis.get().await?;
    let k = key(user_id);
    let items: Vec<String> = conn.zrange(&k, 0, -1).await?;
    let events: Vec<Value> = items
        .iter()
        .filter_map(|item| serde_json::from_str(item).ok())
        .collect();
    Ok(events)
}

/// Clear all pending events for a user (after successful delivery).
pub async fn clear_pending_events(redis: &Pool, user_id: &str) -> Result<(), anyhow::Error> {
    let mut conn = redis.get().await?;
    conn.del::<_, ()>(&key(user_id)).await?;
    Ok(())
}

File: apps/server/src/lib/pending-events.ts
import { redis } from "../db/redis.js";
import type { WSServerEvent } from "@arinova/shared/types";

const KEY_PREFIX = "pending_ws_events:";
const MAX_EVENTS_PER_USER = 1000;
const TTL_SECONDS = 86400; // 24 hours

function key(userId: string) {
  return `${KEY_PREFIX}${userId}`;
}

/**
 * Push a WS event to a user's pending queue.
 * Uses timestamp as score for ordering.
 */
export async function pushEvent(userId: string, event: WSServerEvent) {
  const k = key(userId);
  const score = Date.now();
  await redis.zadd(k, score, JSON.stringify(event));

  // Cap at MAX_EVENTS_PER_USER (remove oldest)
  const count = await redis.zcard(k);
  if (count > MAX_EVENTS_PER_USER) {
    await redis.zremrangebyrank(k, 0, count - MAX_EVENTS_PER_USER - 1);
  }

  // Reset TTL
  await redis.expire(k, TTL_SECONDS);
}

/**
 * Get all pending events for a user.
 */
export async function getPendingEvents(userId: string): Promise<WSServerEvent[]> {
  const k = key(userId);
  const items = await redis.zrange(k, 0, -1);
  return items.map((item) => JSON.parse(item) as WSServerEvent);
}

/**
 * Clear all pending events for a user (after successful delivery).
 */
export async function clearPendingEvents(userId: string) {
  await redis.del(key(userId));
}
